File: Js_Solutions/dijkstra/weighted_graph.js
// Implement Dijkstra's shortest path algorithm

class Node {
    constructor(val, priority) {
        this.value = val;
        this.priority = priority;
    }
}

class PriorityQueue {
    constructor() {
        this.values = [];
    }

    enqueue(val, priority) {
        const newNode = new Node(val, priority);
        this.values.push(newNode);
        this.bubbleUp();
        return true;
    }

    bubbleUp() {
        let idx = this.values.length - 1;
        const element = this.values[idx];
        while (idx > 0) {
            let parentIdx = Math.floor((idx - 1) / 2);
            let parent = this.values[parentIdx];
            if (parent.priority < element.priority) {
                break;
            }
            this.values[idx] = this.values[parentIdx];
            this.values[parentIdx] = element;
            idx = parentIdx;
        }
    }

    dequeue() {
        const min_val = this.values[0];
        const last = this.values.pop();
        if (this.values.length > 0) {
            this.values[0] = last;
            this.sinkDown();
        }
        return min_val;
    }

    sinkDown() {
        let idx = 0;
        const element = this.values[idx];
        const length = this.values.length;

        while (true) {
            let leftChildIdx = (2 * idx) + 1;
            let rightChildIdx = (2 * idx) + 2;
            let leftChild, rightChild;
            let swap = null;

            if (leftChildIdx < length) {
                leftChild = this.values[leftChildIdx];
                if (leftChild.priority < element.priority) {
                    swap = leftChildIdx;
                }
            }

            if (rightChildIdx < length) {
                rightChild = this.values[rightChildIdx];
                if ((swap === null && rightChild.priority < element.priority) || (swap !== null && rightChild.priority < leftChild.priority)) {
                    swap = rightChildIdx;
                }
            }

            if (swap === null) {
                break;
            }
            this.values[idx] = this.values[swap];
            this.values[swap] = element;
            idx = swap;
        }
    }
}

class WeightedGraph {
    constructor() {
        this.adjacencyList = {};
    }

    addVertex(vertex) {
        if (!this.adjacencyList[vertex]) {
            this.adjacencyList[vertex] = [];
        }
    }

    addEdge(vertex1, vertex2, weight) {
        this.adjacencyList[vertex1].push({ node: vertex2, weight });
        this.adjacencyList[vertex2].push({ node: vertex1, weight });
    }

    removeEdge(vertex1, vertex2) {
        let data1 = [];
        let data2 = [];
        for (let neighbor of this.adjacencyList[vertex1]) {
            if (neighbor !== vertex2) {
                data1.push(neighbor);
            }
        }
        this.adjacencyList[vertex1] = data1;
        for (let neighbor of this.adjacencyList[vertex2]) {
            if (neighbor !== vertex1) {
                data2.push(neighbor);
            }
        }
        this.adjacencyList[vertex2] = data2;
    }

    removeVertex(vertex) {
        for (let neighbor of this.adjacencyList[vertex]) {
            this.removeEdge(neighbor, vertex);
        }
        delete this.adjacencyList[vertex];
    }

    Dijkstra(start, finish) {
        const nodes = new PriorityQueue();
        const distances = {};
        const previous = {};
        let smallest;
        const path = []; // to return at the end

        // Build up initial state
        for (let vertex in this.adjacencyList) {
            if (vertex === start) {
                distances[vertex] = 0;
                nodes.enqueue(vertex, 0);
            } else {
                distances[vertex] = Infinity;
                nodes.enqueue(vertex, Infinity);
            }
            previous[vertex] = null;
        }

        // Loop until there is something to visit
        while (nodes.values.length > 0) {
            smallest = nodes.dequeue().value;
            if (smallest === finish) {
                while (previous[smallest]) {
                    path.push(smallest);
                    smallest = previous[smallest]
                }
                break;
            }
            if (smallest || distances[smallest] !== Infinity) {
                for (let neighbor of this.adjacencyList[smallest]) {
                    // Calculate new distance to neighboring node
                    let candidate = distances[smallest] + neighbor.weight;
                    if (candidate < distances[neighbor.node]) {
                        // updating new smallest distance to neighbor
                        distances[neighbor.node] = candidate;
                        // updating previous - How we got to neighbor
                        previous[neighbor.node] = smallest;
                        // Enqueue in priority queue with new priority
                        nodes.enqueue(neighbor.node, candidate);
                    }
                }
            }

        }
        return path.concat(smallest).reverse()
    }

}

g = new WeightedGraph();
g.addVertex('A');
g.addVertex('B');
g.addVertex('C');
g.addVertex('D');
g.addVertex('E');
g.addVertex('F');
g.addEdge('A', 'B', 4);
g.addEdge('A', 'C', 2);
g.addEdge('B', 'E', 3);
g.addEdge('C', 'D', 2);
g.addEdge('C', 'F', 4);
g.addEdge('D', 'E', 3);
g.addEdge('D', 'F', 1);
g.addEdge('E', 'F', 1);
console.log(g.Dijkstra("A", "E"));

File: Python_Solutions/dijkstra/weighted_graph.py
# Implement Dijkstra's shortest path algorithm

class Node:
    def __init__(self, val, priority):
        self.value = val
        self.priority = priority


class PriorityQueue:
    def __init__(self):
        self.values = []

    def enqueue(self, val, priority):
        newNode = Node(val, priority)
        self.values.append(newNode)
        self.bubbleUp()
        return self.values

    def bubbleUp(self):
        idx = len(self.values) - 1
        element = self.values[idx]
        while idx > 0:
            parentIdx = (idx-1)//2
            parent = self.values[parentIdx]
            if parent.priority < element.priority:
                break
            self.values[idx] = self.values[parentIdx]
            self.values[parentIdx] = element
            idx = parentIdx

    def dequeue(self):
        min_val = self.values[0]
        last = self.values.pop()
        if len(self.values) > 0:
            self.values[0] = last
            self.sinkDown()
        return min_val

    def sinkDown(self):
        idx = 0
        element = self.values[idx]
        length = len(self.values)

        while True:
            leftChildIdx = (2*idx) + 1
            rightChildIdx = (2*idx) + 2
            leftChild = None
            rightChild = None
            swap = None

            if leftChildIdx < length:
                leftChild = self.values[leftChildIdx]
                if leftChild.priority < element.priority:
                    swap = leftChildIdx

            if rightChildIdx < length:
                rightChild = self.values[rightChildIdx]
                if (swap is None and rightChild.priority < element.priority) or (swap is not None and rightChild.priority < leftChild.priority):
                    swap = rightChildIdx

            if swap is None:
                break
            self.values[idx] = self.values[swap]
            self.values[swap] = element
            idx = swap


class WeightedGraph:
    def __init__(self):
        self.adjacencyList = {}

    def addVertex(self, vertex):
        if self.adjacencyList.get(vertex) is None:
            self.adjacencyList[vertex] = []

    def addEdge(self, vertex1, vertex2, weight):
        self.adjacencyList[vertex1].append({'node': vertex2, 'weight': weight})
        self.adjacencyList[vertex2].append({'node': vertex1, 'weight': weight})

    def Dijkstra(self, start, finish):
        distances = {}
        previous = {}
        nodes = PriorityQueue()
        path = []  # To return at the end
        smallest = None

        # Setting up initial values
        for vertex in self.adjacencyList.keys():
            if vertex == start:
                distances[vertex] = 0
                nodes.enqueue(vertex, 0)
            else:
                distances[vertex] = float('inf')
                nodes.enqueue(vertex, float('inf'))
            previous[vertex] = None

        # Loop until there is something to visit
        while len(nodes.values) > 0:
            smallest = nodes.dequeue().value
            if smallest == finish:
                while previous[smallest] is not None:
                    path.append(smallest)
                    smallest = previous[smallest]
                break
            if (smallest is not None) or distances[smallest] != float('inf'):
                for neighbor in self.adjacencyList[smallest]:
                    # Calculate new distance to neighbor node
                    candidate = distances[smallest] + neighbor['weight']
                    if candidate < distances[neighbor['node']]:
                        # update new distance to neighbor node
                        distances[neighbor['node']] = candidate
                        # update previous - How we got to neighbor
                        previous[neighbor['node']] = smallest
                        # Enqueue in priority queue with new priority
                        nodes.enqueue(neighbor['node'], candidate)

        path.append(smallest)
        return list(reversed(path))


g = WeightedGraph()
g.addVertex('A')
g.addVertex('B')
g.addVertex('C')
g.addVertex('D')
g.addVertex('E')
g.addVertex('F')
g.addEdge('A', 'B', 4)
g.addEdge('A', 'C', 2)
g.addEdge('B', 'E', 3)
g.addEdge('C', 'D', 2)
g.addEdge('C', 'F', 4)
g.addEdge('D', 'E', 3)
g.addEdge('D', 'F', 1)
g.addEdge('E', 'F', 1)
print(g.Dijkstra("A", "E"))
